File: automation-packages/e2e-testing-playwright/src/main/java/step/examples/loadtesting/playwright/WebmailPlaywrightKeywords.java
package step.examples.loadtesting.playwright;

import com.microsoft.playwright.*;
import com.microsoft.playwright.options.LoadState;
import step.handlers.javahandler.AbstractKeyword;
import step.handlers.javahandler.Keyword;

public class WebmailPlaywrightKeywords extends AbstractKeyword {

    @Keyword(name = "Read order confirmation in Webmail")
    public void readOrderConfirmationInWebmail() {
        try (PlaywrightWrapper playwrightWrapper = PlaywrightWrapper.create()){
            Browser browser = playwrightWrapper.playwright.chromium().launch(new BrowserType.LaunchOptions().setHeadless(false));
            Page page = browser.newPage();
            page.navigate("https://demo-webmail.exense.ch/");
            page.waitForLoadState(LoadState.DOMCONTENTLOADED);
            page.locator("#rcmloginuser").fill("customer@opencart.demo");
            page.locator("#rcmloginpwd").fill("8Fm#%GzdSocv3o");
            page.locator("#rcmloginsubmit").click();

            page.locator("#messagelist").isVisible();
            page.locator("#rcmcountdisplay").isVisible();

            // Click the first email sent by "Your Store"
            page.locator("text=Your Store").nth(0).click();

            FrameLocator iframe = page.frameLocator("#messagecontframe");
            String title = iframe.locator("//h2[@class='subject']").innerText();
            String date = iframe.locator("//div[@class='header-summary']//span[@class='text-nowrap']").innerText();
            //Date dateParsed = new SimpleDateFormat("yyyy-MM-dd HH:mm").parse(date);

            output.add("EmailTitle", title);
            output.add("EmailDate", date);
        }
    }
}

File: automation-packages/e2e-testing-playwright-dotnet/dotnet-keywords/WebmailPlaywrightKeywords.cs
using System.Threading.Tasks;
using Microsoft.Playwright;
using Newtonsoft.Json.Linq;
using Step.Handlers.NetHandler;

namespace Step.Examples.E2ETesting.Playwright
{
    public class WebmailPlaywrightKeywords : AbstractKeyword
    {
        private const string WebmailUrl = "https://demo-webmail.exense.ch/";

        [Keyword(name = "Read order confirmation in Webmail")]
        public void ReadOrderConfirmationInWebmail()
        {
            bool headless = input.Value<bool?>("Headless") ?? true;

            PlaywrightBrowsers.EnsureReady();


            ReadOrderConfirmationAsync(headless).GetAwaiter().GetResult();
        }

        private async Task ReadOrderConfirmationAsync(bool headless)
        {
            using var playwright = await Microsoft.Playwright.Playwright.CreateAsync();
            await using var browser = await playwright.Chromium.LaunchAsync(new BrowserTypeLaunchOptions { Headless = headless });
            var page = await browser.NewPageAsync();

            await page.GotoAsync(WebmailUrl);
            await page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
            await page.Locator("#rcmloginuser").FillAsync("customer@opencart.demo");
            await page.Locator("#rcmloginpwd").FillAsync("8Fm#%GzdSocv3o");
            await page.Locator("#rcmloginsubmit").ClickAsync();

            await page.Locator("#messagelist").WaitForAsync();
            await page.Locator("#rcmcountdisplay").WaitForAsync();

            // Click the first email sent by "Your Store"
            await page.Locator("text=Your Store").Nth(0).ClickAsync();

            var iframe = page.FrameLocator("#messagecontframe");
            string title = await iframe.Locator("//h2[@class='subject']").InnerTextAsync();
            string date = await iframe.Locator("//div[@class='header-summary']//span[@class='text-nowrap']").InnerTextAsync();

            output.Add("EmailTitle", title);
            output.Add("EmailDate", date);
        }
    }
}